### DISCUSSION TITLE
Recommended way of doing user settings with WindowsAppSDK 1.0?

### DISCUSSION THREAD

**Discussion by fourdragons** [microsoft/WindowsAppSDK#1478] [Q&A] [answered]
ApplicationSettingsBase isn't found in System.Configuration, so what is the new way of doing app and user settings with WinUI 3? Thanks!

Edit: I did find the [NuGet package](https://www.nuget.org/packages/System.Configuration.ConfigurationManager/6.0.0-rc.1.21451.13) for using the old way. This still begs the question of is there a new way, or we just install the NuGet package and be on our merry way?

**Comment by andrewleader** [comment, marked as answer]
Hey @fourdragons, thanks for asking!

ApplicationSettingsBase is for WinForms apps. **For WinUI 3 packaged (MSIX) apps, app and user settings should be stored using the ApplicationData class.** Here's instructions on how to save and load user settings: https://docs.microsoft.com/windows/uwp/get-started/settings-learning-track

Note that those docs do say "UWP", however, those docs are applicable to any packaged (MSIX) app, like WinUI 3 packaged.

**Reply by fourdragons** [reply]
All switched over to ApplicationData. Thanks! (It is confusing sometimes that WinUI is UWP... wait it's not... wait it is...) 😁

**Reply by andrewleader** [reply]
Haha... yeah, I hear you 😜 we're working on lots of doc updates, but there's a lot of docs. We've got an awesome doc team that have already started making some improvements to the docs to account for the fact that many APIs are callable from many different types of apps :)

**Reply by DFFspace** [reply]
Saw this thread and wanted to share my way on how I got around to make use of the settings.settings in WinUI. 

I have a Monitoring App that was build as WPF in .NET Framework 4.7 and when .NET 6 got out, I migrated my project to .NET 6. Than when WASDK 1.0.1 got released. I ported my project, mostly copied all files and code in to a new Blank App WinUI 3 desktop project. I had to make some small changes in the XAML code, everything worked fine except for the app settings that I could change and save. At that moment I had no idea how to get around it, so what I did was I copied the: Resource.Designer.cs, Resources.resx and the Settings.settings files from the .NET 6 WPF project in to the WinUI 3 desktop project and that seemed to work. I can open the Settings.settings file and add or remove settings. Launching the App I can launch my settings windows page to change settings and hit the save button to save the settings.

**Reply by DrusTheAxe** [reply]
>For ~WinUI 3~ packaged (MSIX) apps, app and user settings should be stored using the ApplicationData class

FYI ApplicationData is available to all packaged apps, not just those using WinUI3.

**Reply by ThierryFourniol** [reply]
What about unpackaged (non-MSIX) apps? Thanks.

**Reply by riverar** [reply]
> What about unpackaged (non-MSIX) apps? Thanks.

That works too; I use it in WPF. But be warned, this API is very unreliable so be sure to have backup persistence (e.g. registry). Or just use something else. (The roaming feature set of ApplicationData was recently killed removing a bunch of the value from this API.)

**Reply by DrusTheAxe** [reply]
> But be warned, this API is very unreliable

Which API?

**Reply by riverar** [reply]
@DrusTheAxe Windows.Storage.ApplicationData.Current and [...].LocalSettings specifically. I strongly recommend anyone using these properties to wrap access in a try/catch and implement a backup option.

Our app's telemetry is sampled and capped (Bugsnag) but looking at the dashboard today, I'm seeing about 45.6K failures on that API in a window of 30 days.

**Reply by MolotovCherry** [reply]
For anyone coming from Google, if you use an unpackaged app, `Windows.Storage.ApplicationData.Current.LocalSettings` will not work. For unpackaged apps, I used [Config.Net](https://github.com/aloneguid/config) like so:

```c#
public interface IAppSettings : INotifyPropertyChanged {
    // your settings properties here, see Config.Net Github README
}

public static class SettingsProvider {
    public static IAppSettings Initialize() {
        var jsonPath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "my_app_folder", "appsettings.json");

        return new ConfigurationBuilder<IAppSettings>()
           .UseJsonFile(jsonPath)
           .Build();
    }
}
```
Edit: Regarding below replies, to anyone reading this `Environment.SpecialFolder.LocalApplicationData` works too

**Reply by DrusTheAxe** [reply]
`ApplicationData` only exists for packages. Unpackaged apps should use classic Win32 data storage options e.g. `%APPDATALOCAL%\publisher\product`

See https://github.com/microsoft/WindowsAppSDK/issues/2639 for more details

**Reply by riverar** [reply]
Some additional clarity:
* Packaged applications running in AppContainer should use `Windows.Storage.ApplicationData`
* Packaged applications not running in AppContainer (e.g., runFullTrust) can use `Windows.Storage.ApplicationDataManager` or classical methods
* Unpackaged applications should stick to classical methods
